### PR TITLE
python: use --without-computed-gotos to mitigate performance issue

### PR DESCRIPTION
**Pull request by ilovezfs**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

See upstream bug report from 22 Jan 2018 "Significant performance problems with Python 2.7 built with clang 3.x or 4.x"

https://bugs.python.org/issue32616
https://github.com/Homebrew/homebrew-core/issues/22743

### CODE REVIEW

File: Formula/python.rb [hunk 1]
```diff
@@ -103,6 +103,7 @@ def install
       --datadir=#{share}
       --enable-framework=#{frameworks}
       --without-ensurepip
+      --with-computed-gotos
```
Comment (bartskowron) [inline Formula/python.rb]: it should be `--without-computed-gotos`

**Comment by fxcoudert** [comment]
Do we really want to mess with default options to optimise one test case? Do we have any indications that this is systematic, i.e. it will not degrade other benchmarks or performance indicators?

**Comment by ilovezfs** [comment]
Yes. Did you read the upstream ticket?

**Comment by fxcoudert** [comment]
I did, and it didn't say that upstream was considering making this move, nor that it had been tested systematically (actually, it did say it had not been tested systematically). I tried to check the options that Apple python is compiled with, but I cannot find trace of `--without-computed-gotos` in `sysconfig.get_config_vars()` (but maybe it's the wrong place to look).

**Comment by ilovezfs** [comment]
I think they're not compiling their Python with Xcode 9.

**Comment by ilovezfs** [comment]
@fxcoudert in any case this was recommended by @methane who is from upstream.

@methane do you have any thoughts on @fxcoudert's concerns?

**Comment by ilovezfs** [comment]
@zmwangx do you mind re-running your benchmarks for comparison with the `--without-computed-gotos` in there?

**Comment by zmwangx** [comment]
I'll try to run systematic benchmarks tonight (going to take hours).

**Comment by zmwangx** [comment]
By the way, here's the compiler version used to compile system Python (not sure if that clang is some experimental build):

```console
$ /usr/bin/python -c 'import sys; print(sys.version)'
2.7.10 (default, Jul 15 2017, 17:16:57)
[GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.31)]
```

**Comment by zmwangx** [comment]
Also, there might be clues in [1]; I just don't have to check lately.

[1] https://opensource.apple.com/source/python/python-109/

**Comment by ilovezfs** [comment]
That looks like run of the mill Xcode 9 to me … :(

**Comment by ilovezfs** [comment]
I built 2.7.10 from the old brew formula in b27c68774d76ae7d269864e3937e3f10b7d3eebd and lo and behold there is no issue:
```
$ /usr/local/bin/python brewpythonsucks.py 
11.2221500874
$ /usr/bin/python brewpythonsucks.py 
10.2731819153
```
So that would explain why system Python doesn't have the problem.

**Comment by ilovezfs** [comment]
Given this seems to be a regression in Python itself, I'm going to close this out.

**Comment by zmwangx** [comment]
Turns out Apple made the "right" choice sticking to 2.7.10; the performance hit started in 2.7.11.

**Comment by ilovezfs** [comment]
Interesting. Have you begun bisecting 2.7.10 -> 2.7.11?

**Comment by zmwangx** [comment]
Yeah, on it.

**Comment by ilovezfs** [comment]
nice

**Comment by zmwangx** [comment]
@ilovezfs Plot twist: this PR can be reopened. The problematic commit is precisely https://github.com/python/cpython/commit/2c992a0788536087bfd78da8f2c62b30a461d7e2:

```
commit 2c992a0788536087bfd78da8f2c62b30a461d7e2
Author: Benjamin Peterson <benjamin@python.org>
Date:   Thu May 28 12:45:31 2015 -0500

    backport computed gotos (#4753)
```

```console
$ /tmp/python-build/78ba242403/bin/python /tmp/test.py
1.864752
$ /tmp/python-build/2c992a0788/bin/python /tmp/test.py
7.906417
```

**Comment by ilovezfs** [comment]
Wow. Nice sleuthing @zmwangx. Maybe upstream can automatically disable that for the affected compiler versions.

**Comment by ilovezfs** [comment]
@zmwangx so that still leaves open the question why this isn't a problem for Python 3.x. Were further changes made to computed gotos stuff in Python 3.x that haven't been backported?

**Review by bxsx — COMMENTED** [review]
Ok, this is a newbie question. Why these vim, sip and macvim formulas are changed in this PR? What relation is here?

**Comment by ilovezfs** [comment]
@bartskowron those formulae have /usr/local/Cellar/python/... paths hard coded so they break whenever the realpath to the installed package changes.

**Comment by bartskowron** [comment]
I believe there was some important arguments for hardcoding here. Cause it's hard to track which exactly formulas are affected and require updating revisions. Why these 3 and no any others?

**Review by fxcoudert — APPROVED** [review]

**Comment by bartskowron** [comment]
Ok. As a homebrew's user I would say lets go with the patch.
However, as @ilovezfs noticed it seems like some other changes made in 
Python3 were not backported. I don't know the current PSF policy regarding Python 2.x. Are they still going to release another Python2 version with fixes? If so, this issue should be fixed in the upstream to not keep a temporary special case here. However, I remember (but cannot confirm) that the 2.7.14 was supposed to be the last release? If so, it's on the homebrew side to fix it even if temporary.

**Review by DomT4 — APPROVED** [review]

**Comment by zmwangx** [comment]
@ilovezfs 

> so that still leaves open the question why this isn't a problem for Python 3.x. Were further changes made to computed gotos stuff in Python 3.x that haven't been backported?

Not sure, computed gotos has been in the Python 3 branch since the py3k days and was first included in v3.1. See https://github.com/python/cpython/commit/b52ec78baf236272800017fdb2b7f243741c1851. [CPython core developer](https://bugs.python.org/issue32616#msg311751) is puzzled by this too.

@bartekrutkowski 

> Are they still going to release another Python2 version with fixes? If so, this issue should be fixed in the upstream to not keep a temporary special case here. However, I remember (but cannot confirm) that the 2.7.14 was supposed to be the last release? If so, it's on the homebrew side to fix it even if temporary.

As a downstream packager we can enable non-default flags or even carry long term patches at our discretion. That's not really a problem. Also, Python 2.7 still has two years of lifespan, and there will be a 2.7.15 in 2018. See [PEP 373](https://www.python.org/dev/peps/pep-0373/).

**Comment by methane** [comment]
> this issue should be fixed in the upstream 

I created pull request: python/cpython#5574

**Comment by fxcoudert** [comment]
Thanks everyone involved! Quite a success story :)

**Pull request closed without merging**